components/org.wso2.carbon.identity.sp.metadata.saml2/src/main/java/org/wso2/carbon/identity/sp/metadata/saml2/util/Parser.java

```diff
@@ -245,21 +245,61 @@ public SAMLSSOServiceProviderDO parse(String metadata, SAMLSSOServiceProviderDO
      */
     private EntityDescriptor generateMetadataObjectFromString(String metadataString) {
         EntityDescriptor entityDescriptor = null;
+        InputStream inputStream;
         try {
-            DocumentBuilderFactory factory = IdentityUtil.getSecuredDocumentBuilderFactory();
-            factory.setNamespaceAware(true);
-            DocumentBuilder builder = factory.newDocumentBuilder();
-            Document document = builder.parse(new ByteArrayInputStream(metadataString.getBytes()));
-            Element node = document.getDocumentElement();
-            DOMMetadataProvider idpMetaDataProvider = new DOMMetadataProvider(node);
-            idpMetaDataProvider.setRequireValidMetadata(true);
-            idpMetaDataProvider.setParserPool(new BasicParserPool());
-            idpMetaDataProvider.initialize();
-            XMLObject xmlObject = idpMetaDataProvider.getMetadata();
-            entityDescriptor = (EntityDescriptor) xmlObject;
-        } catch (MetadataProviderException | SAXException | ParserConfigurationException | IOException e) {
+            doBootstrap();
+            inputStream = new ByteArrayInputStream(metadataString.trim().getBytes(StandardCharsets.UTF_8));
+            entityDescriptor = (EntityDescriptor) XMLObjectSupport.unmarshallFromInputStream(
+                    XMLObjectProviderRegistrySupport.getParserPool(), inputStream);
+        } catch ( UnmarshallingException | XMLParserException e) {
             log.error("Error While reading Service Provider metadata xml", e);
         }
         return entityDescriptor;
     }
+
+    public static void doBootstrap() {
```

Comment:
let's have a single place with bootstrap logic, rather than duplicating the code

Reply:
Initialization logic is centralized at org.wso2.carbon.identity.core.util with the class named **SAMLInitializer**